### DISCUSSION TITLE
MacOS: can't access local_socket created in a process running as root from a non-root process.

### DISCUSSION THREAD

**Discussion by radiohertz** [kotauskas/interprocess#65] [Q&A]
The listener is created by a process running as root and the client is a non-root process. 

The client fails to connect with 
```
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

My only guess is that the client and server are generating two different names. 
The server creates a socket in `/tmp/sock` and when i try to connect to the server with the same name `sock`, it can't find the file.

# Example 

```rs
/// src/main.rs

use std::io::Read;

use interprocess::{
    local_socket::{traits::Listener, ListenerOptions, ToNsName},
    os::unix::local_socket::SpecialDirUdSocket,
};

fn main() {
    let listener = ListenerOptions::new()
        .name("test".to_ns_name::<SpecialDirUdSocket>().unwrap())
        .create_sync()
        .unwrap();

    loop {
        let mut stream = listener.accept().unwrap();
        println!("got a connection");
        let mut buf = [0; 1024];
        if let Err(e) = stream.read(&mut buf) {
            println!("{e}");
        }
    }
}

```

```rs
/// examples/client.rs

use interprocess::{
    local_socket::{self, traits::Stream, ToNsName},
    os::unix::local_socket::SpecialDirUdSocket,
};

fn main() {
    let stream =
        local_socket::Stream::connect("test".to_ns_name::<SpecialDirUdSocket>().unwrap()).unwrap();
}

```

**Comment by kotauskas** [comment]
> My only guess is that the client and server are generating two different names.

[That is not possible.](https://github.com/kotauskas/interprocess/blob/2.1.0/src/os/unix/local_socket/name_type.rs#L61-L74)

If you get a directory listing for `/tmp` from the command line (`ls -lh /tmp`), does it contain the socket file? If not, does it contain it if `ls` is run as `root`?

**Reply by radiohertz** [reply]
It does create the socket in `/tmp` when i run the server as root but i dont see it in `/tmp` when it's not root.

**Reply by radiohertz** [reply]
And, when i run the server the second time as non-root, it fails obviously because the socket is already there but idk where to find the socket to delete it.

**Reply by kotauskas** [reply]
That makes it a permissions issue, which can be solved via [`interprocess::os::unix::local_socket::ListenerOptionsExt::mode()`](<https://docs.rs/interprocess/2.1.0/interprocess/os/unix/local_socket/trait.ListenerOptionsExt.html#tymethod.mode>). Note that you must only make this call on Unix systems and avoid it altogether on Windows using conditional compilation, as your crate wouldn't compile on Windows if it is made unconditionally.

Regarding the corpse socket you're seeing when running as non-`root`, that is supposed to be automatically removed by Interprocess if the server is shut down gracefully, via [name reclamation](<https://docs.rs/interprocess/2.1.0/interprocess/local_socket/enum.Listener.html#name-reclamation>). Since your example is so simple, there isn't a graceful shutdown mechanism, and so the server never ends up cleaning up the corpse socket, causing the non-`root` run to fail.

To get rid of the corpse socket manually, run `rm` as `root`. It's still in `/tmp`, but insufficiently privileged access attempts seem to be returning "not found" instead of "access denied" for some reason. This error misreporting I cannot help you with – this is macOS magic I'm not aware of (because I don't use macOS and don't have a macOS system to test this on aside from GitHub Actions)

**Reply by radiohertz** [reply]
> That makes it a permissions issue, which can be solved via [interprocess::os::unix::local_socket::ListenerOptionsExt::mode()](https://docs.rs/interprocess/2.1.0/interprocess/os/unix/local_socket/trait.ListenerOptionsExt.html#tymethod.mode).

Which mode should i be using? I see that the default is 666, which means it should be readable and writable for everyone right?

**Reply by kotauskas** [reply]
That is true, but there might be macOS-specific oddities preventing the mode from being 666₈. What's the mode of the socket file if you try and query it as `root` (using `ls` or `stat`)?

**Reply by radiohertz** [reply]
It's `srwxr-xr-x`. 

Very weirdly, it's shows ```called `Result::unwrap()` on an `Err` value: Os { code: 61, kind: ConnectionRefused, message: "Connection refused" }``` now. It was showing `File Not Found` before, very interesting.

**Reply by kotauskas** [reply]
The "connection refused" error is indicative of a corpse socket and is the typical error seen when trying to connect to a socket created by a server that didn't get the chance to clean it up.

The cause for the `umask` not applying seems to be [a bug of my own creation](https://github.com/kotauskas/interprocess/blob/2.1.0/src/os/unix/c_wrappers.rs#L230) which I will address shortly. As a temporary fix, you can use [`libc::umask(0o000)`](https://docs.rs/libc/latest/libc/fn.umask.html) directly.

**Reply by radiohertz** [reply]
> As a temporary fix, you can use [libc::umask(0o666)](https://docs.rs/libc/latest/libc/fn.umask.html) directly.

Still seems to behave the same. Here's what i did.

```rs
    let old = unsafe { libc::umask(0o666) };
    let listener = ListenerOptions::new()
        .name("test1".to_ns_name::<SpecialDirUdSocket>().unwrap())
        .create_sync()
        .unwrap();
    unsafe { libc::umask(old) };
```
Still stuck with `No such file or directory`. The permissions are `s--x--x--x`.

**Reply by kotauskas** [reply]
Apologies for the mixup, I'm just not very used to `umask()` working the opposite way of how its name suggests it should work. The correct call is `libc::umask(0o000)`.

**Reply by radiohertz** [reply]
It seems to have the correct permissions now `srwxrwxrwx` but im still getting `No such file or directory` when the client is non root.

**Reply by kotauskas** [reply]
What are the permissions of `/tmp`'s destination (it's reportedly a symlink on macOS)? The OS [may have screwed them up](https://superuser.com/a/775889/754580) for one reason or another.

**Reply by radiohertz** [reply]
it's `lrwxr-xr-x@` and it does symlink to `/private/tmp`.

**Reply by kotauskas** [reply]
In that case, you should `sudo chmod 1777 /tmp` and the same for `/private/tmp`, as those are the standard permissions for `/tmp`, and are expected by portable software written for Unix-like systems.

**Reply by radiohertz** [reply]
I did change the perms on the directories and it's still `No such file or directory` and i'm just tired at this point. Anyway, thanks for helping out.

**Reply by kotauskas** [reply]
You're welcome!

I'm bothered by macOS's odd behavior here. `SpecialDirUdSocket` isn't supposed to fail like this, as it expects the directory it points to to be world-accessible, ideally with the sticky bit (so `1777`). While I can't easily provide a better solution without introducing a breaking change, I'll try my best to document this issue in an approachable way.
